writeaheadlog_silo_test.go

```diff
@@ -0,0 +1,332 @@
+	binary.LittleEndian.PutUint32(data[:], su.number)
+	_, err := silo.f.WriteAt(data[:], su.offset)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// threadedSetupWrite simulate a setup by writing random data to disk
+func (s *silo) threadedSetupWrite(done chan error, dataPath string) {
+	// signal completion
+	defer close(done)
+
+	// delete old data file if it already exists
+	err := os.Remove(filepath.Join(dataPath, hex.EncodeToString(s.cs[:])))
+	if err != nil && !os.IsNotExist(err) {
+		done <- err
+		return
+	}
+
+	// write checksum
+	s.updateChecksum()
+	_, err = s.f.WriteAt(s.cs[:], s.offset+int64(len(s.numbers)*4))
+	if err != nil {
+		done <- err
+		return
+	}
+
+	// write new data file
+	newFile, err := os.Create(filepath.Join(dataPath, hex.EncodeToString(s.cs[:])))
+	if err != nil {
+		done <- err
+		return
+	}
+	_, err = newFile.Write(fastrand.Bytes(pageSize))
+	if err != nil {
+		done <- err
+		return
+	}
+	if err := newFile.Close(); err != nil {
+		done <- err
+		return
+	}
+
+	// sync changes
+	newFile.Sync()
+	s.f.Sync()
+}
+
+// threadedUpdate updates a silo until stopChan is closed. This leaves the last
+// transaction unfinished and therefore corrupt.
+func (s *silo) threadedUpdate(t *testing.T, w *WAL, dataPath string, wg *sync.WaitGroup) {
+	defer wg.Done()
+
+	sus := make([]siloUpdate, 0, len(s.numbers))
+	updates := make([]Update, 0, len(s.numbers))
+	iterations := 1000
+	for i := 0; i < iterations; i++ {
+		// Change between 1 and len(s.numbers)
+		length := rand.Intn(len(s.numbers)) + 1
+		for j := 0; j < length; j++ {
+			if s.i == 0 {
+				s.numbers[s.i] = (s.numbers[len(s.numbers)-1] + 1)
+			} else {
+				s.numbers[s.i] = (s.numbers[s.i-1] + 1)
+			}
+			sus = append(sus, s.newSiloUpdate(s.i, s.numbers[s.i]))
+
+			// Increment the index. If that means we reach the end, set it to 0
+			s.i = (s.i + 1) % uint32(len(s.numbers))
+		}
+
+		// Convert silo updates to WAL updates
+		for _, su := range sus {
+			updates = append(updates, newUpdate(su))
+		}
+
+		// Start setup write
+		wait := make(chan error)
+		go s.threadedSetupWrite(wait, dataPath)
+
+		// Create txn
+		txn, err := w.NewTransaction(updates)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		// Wait for setup to finish
+		if err := <-wait; err != nil {
+			t.Fatal(err)
+		}
+
+		// Signal setup complete
+		if err := <-txn.SignalSetupComplete(); err != nil {
+			t.Fatal(err)
+		}
+
+		// Corrupt the last iteration by not finishing it
+		if i == iterations-1 {
+			return
+		}
+
+		// Apply the updates
+		for _, su := range sus {
+			su.applyUpdate(s)
+		}
+
+		// Signal release complete
+		if err := txn.SignalUpdatesApplied(); err != nil {
+			t.Fatal(err)
+		}
+
+		// Reset
+		sus = sus[:0]
+		updates = updates[:0]
+	}
+}
+
+// TestSilo is an integration test that is supposed to test all the features of
+// the WAL in a single testcase. It uses 100 silos updating 1000 times each.
+func TestSilo(t *testing.T) {
+	testdir := build.TempDir("wal", t.Name())
```

Comment:
in a separate PR we should replace all uses of `build.TempDir`, because it puts things in a `SiaTesting` directory

Reply:
Yes, we should remove all the dependencies to the Sia repo.